src/IssueViz.Security.UnitTests/Hotspots/LocalHotspotStoreTests.cs

```diff
@@ -0,0 +1,343 @@
+        eventListener.Events.Should().HaveCount(2).And.Subject.Last().Should()
+            .BeEquivalentTo(new IssuesChangedEventArgs(Array.Empty<IAnalysisIssueVisualization>(), newHotspots));
+    }
+
+    [TestMethod]
+    public void AddOrUpdate_ServerHotspots_MatchesCorrectly()
+    {
+        var serverStoreMock = new Mock<IServerHotspotStore>();
+        var serverHotspot1 = CreateEmptyServerHotspot();
+        var serverHotspot2 = CreateEmptyServerHotspot();
+        serverStoreMock.Setup(x => x.GetAll()).Returns(new[] { serverHotspot1, serverHotspot2 });
+
+        var issueVis3 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis1 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis2 = Mock.Of<IAnalysisIssueVisualization>();
+
+        var matcherMock = new Mock<IHotspotMatcher>();
+        matcherMock.Setup(x => x.IsMatch(issueVis1, serverHotspot1)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis2, serverHotspot2)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis3, serverHotspot1))
+            .Returns(true); // 2 local hotspots match 1 server but only one pair should be created
+
+        var testSubject = CreateTestSubject(out _, serverStoreMock.Object, matcherMock.Object);
+
+        testSubject.AddOrUpdate("lalala", new[] { issueVis1, issueVis2, issueVis3 });
+
+        VerifyContent(testSubject, new LocalHotspot(issueVis1, serverHotspot1),
+            new LocalHotspot(issueVis2, serverHotspot2), new LocalHotspot(issueVis3));
+    }
+
+    [TestMethod]
+    public void AddOrUpdate_ServerHotspots_SameFileUpdate_MakesUnmatchedServerHotspotsAvailable()
+    {
+        var serverStoreMock = new Mock<IServerHotspotStore>();
+        var serverHotspot1 = CreateEmptyServerHotspot();
+        var serverHotspot2 = CreateEmptyServerHotspot();
+        serverStoreMock.Setup(x => x.GetAll()).Returns(new[] { serverHotspot1, serverHotspot2 });
+
+        var issueVis1 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis2 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis3 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis4 = Mock.Of<IAnalysisIssueVisualization>();
+
+        var matcherMock = new Mock<IHotspotMatcher>();
+        matcherMock.Setup(x => x.IsMatch(issueVis1, serverHotspot1)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis2, serverHotspot2)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis3, serverHotspot1)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis4, serverHotspot2)).Returns(true);
+
+        var testSubject = CreateTestSubject(out _, serverStoreMock.Object, matcherMock.Object);
+
+        testSubject.AddOrUpdate("lalala", new[] { issueVis1, issueVis2 });
+        testSubject.AddOrUpdate("lalala", new[] { issueVis3 });
+        testSubject.AddOrUpdate("lalala", new[] { issueVis3, issueVis4 });
```

Comment:
It's hard to follow what this is testing and why the final result is the expected one.
The test is making three state changes to the store but only checking the final one. It would be clearer and more robust if it validate the contents after each event.